Add a CONTRIBUTING.md file

sunfishcode opened the pull request:
Here's an attempt at a CONTRIBUTING.md file, to welcome newcomers and answer questions about how to get involved in the project. I've included answers to questions that I get a lot, but if there are things I've missed that would help people get started, please bring them up!

bnjbvr commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,58 @@
+
+All PRs must be formatted according to the current stable rustfmt, and this is
+checked in the continuous integration tests. See
+[here](https://github.com/rust-lang-nursery/rustfmt#quick-start) for setup.
```
Is it a good place to mention the `format-all.sh` script?

sunfishcode replied:
Mentioned.

sunfishcode commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,58 @@
+compiler backends, learning about machine code, learning about how Cranelift
+does things, or all together at once.
+
+We tag issues in the issue tracker marked "good first issue" when we can, so
```
Good idea.

sunfishcode commented:
> mention ways to get in touch with other contributors, that is, give links to IRC / gitter rooms, or anything (it is probably somewhere else already, but I couldn't find it).

Added.

> not sure whether Cranelift has a Code of Conduct, but if it were the case, it would be nice to mention it here.

We don't have one yet, though I expect to propose one soon, at which point, yes, we should mention it in CONTRIBUTING.md.

> I'd add something about the process to contribute code: does one need to open an issue all the time? Or should PRs just be sent without previous discussion/approval? Are there specific commit message restrictions/recommendations? Is there a "one feature per commit" policy? Should every single commit compile and pass tests (which is practical for bisection purposes)?

> random idea: a word on governance might or might not be appropriate: what's the process to decide whether a PR is accepted or not? i.e. who gets the last word to decide whether a PR is accepted? (it's rather clear in my head, but might be nice to make it explicit)

I've now added a Development Process section which I think covers all these questions. Let me know what you think!

> Are there ways non technical contributions can be made to the project? (documentation for instance, talking about the project, etc. -- makes me wonder whether a list of public talks/slidedecks about Cranelift could find a way somewhere in the repository)

We don't have a lot right now, but it'd be cool to start collecting them.

6A commented on CONTRIBUTING.md:
```diff
@@ -74,3 +74,36 @@ Our Python code is checked with [mypy](http://mypy-lang.org/) and
+[issues]: https://guides.github.com/features/issues/
+[pull requests]: https://help.github.com/articles/about-pull-requests/
+[issue keywords]: https://help.github.com/articles/closing-issues-using-keywords/
+[Core Team]: https://github.com/orgs/CraneStation/teams/core-team
```
I think `/teams` and `/teams/core-team` are restricted to members of the CraneStation organization. I think https://github.com/orgs/CraneStation/people would do fine though.

sunfishcode replied:
Ah, thanks for pointing that out. I've updated it to the people link.

sunfishcode commented:
Thanks for the feedback everyone! I expect this will continue to evolve, but it's a starting point :-).
